File: jre/src/main/java/azure/data/cosmos/serialization/hybridrow/internal/Utf8StringJsonConverter.java
//------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

package azure.data.cosmos.serialization.hybridrow.internal;

import Newtonsoft.Json.*;

/**
 * Helper class for parsing <see cref="Utf8String" /> from JSON.
 */
public class Utf8StringJsonConverter extends JsonConverter {
    @Override
    public boolean getCanWrite() {
        return true;
    }

    @Override
    public boolean CanConvert(java.lang.Class objectType) {
        return objectType.isAssignableFrom(Utf8String.class);
    }

    @Override
    public Object ReadJson(JsonReader reader, java.lang.Class objectType, Object existingValue,
                           JsonSerializer serializer) {
        Contract.Requires(reader.TokenType == JsonToken.String);
        return Utf8String.TranscodeUtf16((String)reader.Value);
    }

    @Override
    public void WriteJson(JsonWriter writer, Object value, JsonSerializer serializer) {
        writer.WriteValue(((Utf8String)value).toString());
    }
}

File: src/Serialization/HybridRow/Internal/Utf8StringJsonConverter.cs
namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Internal
{
    using System;
    using Microsoft.Azure.Cosmos.Core;
    using Microsoft.Azure.Cosmos.Core.Utf8;
    using Newtonsoft.Json;

    /// <summary>Helper class for parsing <see cref="Utf8String" /> from JSON.</summary>
    internal class Utf8StringJsonConverter : JsonConverter
    {
        public override bool CanWrite => true;

        public override bool CanConvert(Type objectType)
        {
            return typeof(Utf8String).IsAssignableFrom(objectType);
        }

        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
        {
            Contract.Requires(reader.TokenType == JsonToken.String);
            return Utf8String.TranscodeUtf16((string)reader.Value);
        }

        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
        {
            writer.WriteValue(((Utf8String)value).ToString());
        }
    }
}
